pkg/controller/apimanager/apimanager_controller_test.go

```diff
@@ -0,0 +1,110 @@
+package apimanager
+
+import (
+	"context"
+	"testing"
+
+	"github.com/3scale/3scale-operator/pkg/3scale/amp/operator"
+	appsv1alpha1 "github.com/3scale/3scale-operator/pkg/apis/apps/v1alpha1"
+	appsv1 "github.com/openshift/api/apps/v1"
+	imagev1 "github.com/openshift/api/image/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func TestAPIManagerControllerCreate(t *testing.T) {
+	var (
+		name           = "example-apimanager"
+		namespace      = "operator-unittest"
+		wildcardDomain = "test.3scale.net"
+	)
+
+	apimanager := &appsv1alpha1.APIManager{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: appsv1alpha1.APIManagerSpec{
+			APIManagerCommonSpec: appsv1alpha1.APIManagerCommonSpec{
+				WildcardDomain: wildcardDomain,
+			},
+		},
+	}
+
+	// Objects to track in the fake client.
+	objs := []runtime.Object{apimanager}
+
+	// Register operator types with the runtime scheme.
+	s := scheme.Scheme
+	s.AddKnownTypes(appsv1alpha1.SchemeGroupVersion, apimanager)
+	err := appsv1.AddToScheme(s)
+	if err != nil {
+		t.Fatalf("Unable to add Apps scheme: (%v)", err)
+	}
+	imagev1.AddToScheme(s)
+	if err != nil {
+		t.Fatalf("Unable to add Image scheme: (%v)", err)
+	}
+	routev1.AddToScheme(s)
+	if err != nil {
+		t.Fatalf("Unable to add Route scheme: (%v)", err)
+	}
+
+	// Create a fake client to mock API calls.
+	cl := fake.NewFakeClient(objs...)
+	clientAPIReader := fake.NewFakeClient(objs...)
+
+	baseReconciler := operator.NewBaseReconciler(cl, clientAPIReader, s, log)
+	baseControllerReconciler := operator.NewBaseControllerReconciler(baseReconciler)
+
+	// Create a ReconcileMemcached object with the scheme and fake client.
+	r := ReconcileAPIManager{
+		BaseControllerReconciler: baseControllerReconciler,
+	}
+
+	req := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+
+	res, err := r.Reconcile(req)
+	if err != nil {
+		t.Fatalf("reconcile: (%v)", err)
+	}
+
+	if !res.Requeue {
+		t.Error("reconcile did not requeue request as expected. Requeuing due to setting of defaults should have been performed")
+	}
+
+	res, err = r.Reconcile(req)
+	if err != nil {
+		t.Fatalf("reconcile: (%v)", err)
+	}
+
+	if res.Requeue {
+		t.Error("reconcile did not finish end of reconciliation as expected. APIManager should have been reconciled at this point")
+	}
+
```

Comment:
If you look up at until this point you will see that I check that the first reconcile call returns a requeue, as we expect to some defaults to be set with the provided apimanager), and that I check that the second call does not return a requeue as it is expected all objects are created successfully.